server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/RegexMappingRoutes.java

```diff
@@ -0,0 +1,128 @@
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+
+import org.apache.james.rrt.api.RecipientRewriteTable;
+import org.apache.james.rrt.api.RecipientRewriteTableException;
+import org.apache.james.rrt.api.RegexInvalidException;
+import org.apache.james.rrt.lib.MappingSource;
+import org.apache.james.webadmin.Constants;
+import org.apache.james.webadmin.Routes;
+import org.apache.james.webadmin.dto.RegexMappingDTO;
+import org.apache.james.webadmin.utils.ErrorResponder;
+import org.apache.james.webadmin.utils.JsonExtractor;
+import org.eclipse.jetty.http.HttpStatus;
+
+import com.google.common.annotations.VisibleForTesting;
+
+import io.swagger.annotations.Api;
+import io.swagger.annotations.ApiOperation;
+import io.swagger.annotations.ApiResponse;
+import io.swagger.annotations.ApiResponses;
+import spark.HaltException;
+import spark.Request;
+import spark.Response;
+import spark.Service;
+
+@Api(tags = "Adding")
+@Path(RegexMappingRoutes.BASE_PATH)
+@Produces(Constants.JSON_CONTENT_TYPE)
+public class RegexMappingRoutes implements Routes {
+
+    static final String BASE_PATH = "/mappings/regex";
+
+    private final RecipientRewriteTable recipientRewriteTable;
+    private final JsonExtractor<RegexMappingDTO> jsonExtractor;
+
+    @Inject
+    @VisibleForTesting
+    RegexMappingRoutes(RecipientRewriteTable recipientRewriteTable) {
+        this.recipientRewriteTable = recipientRewriteTable;
+        this.jsonExtractor = new JsonExtractor<>(RegexMappingDTO.class);
+    }
+
+    @Override
+    public String getBasePath() {
+        return BASE_PATH;
+    }
+
+    @Override
+    public void define(Service service) {
+        service.post(getBasePath(), this::addRegexMapping);
+    }
+
+    @POST
+    @Path(BASE_PATH)
+    @ApiOperation(value = "adding address-regex mappings to RecipientRewriteTable")
+    @ApiResponses(value = {
+        @ApiResponse(code = HttpStatus.NO_CONTENT_204, message = "Mapping successfully added"),
+        @ApiResponse(code = HttpStatus.BAD_REQUEST_400, message = "Invalid body")
+    })
+    public HaltException addRegexMapping(Request request, Response response) throws Exception {
+        RegexMappingDTO regexMappingDTO = jsonExtractor.parse(request.body());
+        MappingSource mappingSource = extractMappingSource(regexMappingDTO);
+        String regex = extractRegex(regexMappingDTO);
+
+        try {
+            recipientRewriteTable.addRegexMapping(mappingSource,regex);
+        } catch (RecipientRewriteTableException e) {
+            throw ErrorResponder.builder()
+                .statusCode(HttpStatus.INTERNAL_SERVER_ERROR_500)
+                .type(ErrorResponder.ErrorType.INVALID_ARGUMENT)
+                .message(e.getMessage())
+                .haltError();
+        }
+        return halt(HttpStatus.NO_CONTENT_204);
+    }
+
+    private MappingSource extractMappingSource(RegexMappingDTO regexMappingDTO) {
+        try {
+            return MappingSource.parse(regexMappingDTO.getSource());
```

Comment:
Rely on MailAddressParser...

Reply:
I have changed to MappingSource.fromUser() method instead of parse(). Maybe it's better ?